packages/python/plotly/plotly/express/_core.py

```diff
@@ -754,6 +754,107 @@ def apply_default_cascade(args):
         args["marginal_x"] = None
 
 
+def build_or_augment_dataframe(args, attrables, array_attrables):
+    """
+    Constructs an implicit dataframe and modifies `args` in-place.
+
+    Parameters
+    ----------
+    args : OrderedDict
+    `attrables` is a list of keys into `args`, all of whose corresponding
+    values are converted into columns of a dataframe.
+    Used to be support calls to plotting function that elide a dataframe
+    argument; for example `scatter(x=[1,2], y=[3,4])`.
+    """
+
+    # We start from an empty DataFrame except for the case of functions which
+    # implicitely need all dimensions: Splom, Parcats, Parcoords
+    # This could be refined when dimensions is given
+    df = pd.DataFrame()
+
+    if "dimensions" in args and args["dimensions"] is None:
+        if args.get("data_frame") is None or args["data_frame"] is None:
+            raise ValueError(
+                "No data were provided. Please provide data either with the `data_frame` or with the `dimensions` argument."
+            )
+        else:
+            df_args = args["data_frame"]
+            df[df_args.columns] = df_args[df_args.columns]
+
+    # Valid column names
+    df_columns = (
+        args["data_frame"].columns if args.get("data_frame") is not None else None
+    )
+    group_attrs = ["symbol", "line_dash"]
+    for group_attr in group_attrs:
+        if group_attr in args:
+            attrables += [group_attr]
+
+    # Loop over possible arguments
+    for field_name in attrables:
+        argument_list = (
+            [args.get(field_name)]
+            if field_name not in array_attrables
+            else args.get(field_name)
+        )
+        if argument_list is None:  # argument not specified, continue
+            continue
+        # Argument name: field_name if the argument is a list
+        # Else we give names like ["hover_data_0, hover_data_1"] etc.
+        field_list = (
+            [field_name]
+            if field_name not in array_attrables
+            else [field_name + "_" + str(i) for i in range(len(argument_list))]
+        )
+        # argument_list and field_list ready, iterate over them
+        for i, (argument, field) in enumerate(zip(argument_list, field_list)):
+            if argument is None:
+                continue
+            elif isinstance(argument, str):  # just a column name
+                # Check validity of column name
+                try:
```

Comment:
what if `args["data_frame"]` isn't provided? we need a special error message for "you can't just give us strings without a dataframe!" or something

Reply:
done!